inspirehep/dojson/hep/fields/bd6xx.py

```diff
@@ -106,37 +77,125 @@ def accelerator_experiments2marc(self, key, value):
     }
 
 
-@hep.over('thesaurus_terms', '^695..')
-def thesaurus_terms(self, key, value):
-    """Controlled keywords."""
-    value = force_force_list(value)
+@hep.over('keywords', '^695..')
+@hep.over('keywords', '^653[10_2][_1032546]')
+@utils.for_each_value
+def keywords(self, key, value):
+    """Parse keywords.
+
+    We have 2 types of keywords from marc, the 'thesaurus_terms' and
+    'freekeys', and we want to merge them for json into a big 'keywords' and
+    extract the 'energy_ranges' too that is represented in marc as special
+    value of some keywords.
+    """
+    def _is_thesaurus(value):
+        return '9' not in value
+
+    def _is_energy(value):
+        return 'e' in value
+
+    def _freekey_get_source(source_value):
+        if not isinstance(source_value, (list, tuple, set)):
+            return source_value
+
+        source_value = [source.lower() for source in source_value]
+        if 'conference' in source_value:
+            return ''
+
+        for source in source_value:
+            if source in ['author', 'publisher']:
+                return source
+
+        return ''
+
+    def _freekey_get_dict(elem):
+        keyword = {
+            'keyword': elem.get('a'),
+            'classification_scheme': '',
+        }
+
+        source = _freekey_get_source(elem.get('9'))
+        if source:
+            keyword['source'] = source
+
+        return keyword
 
-    def get_value(value):
-        try:
-            energy_range = int(value.get('e'))
-        except (TypeError, ValueError):
-            energy_range = None
+    def _get_keyword_dict(elem):
+        if _is_thesaurus(elem) and elem.get('a'):
+            return {
+                'keyword': elem.get('a'),
+                'classification_scheme': elem.get('2'),
+                'source': '',
```

Comment:
I misunderstood it then, I though we said to use the source only for free keywords, and the scheme only for thesaurus, and leave it empty otherwise (so you can distinguish them). If there's a thesaurus with source and no schema, we will be unable to know if it was thesaurus or free keyword, how that should be handled?